File: example/android/device/src/main/java/com/qc/device/model/Device.kt
package com.qc.device.model

data class Device(
    /**
     * 电池
     */
    val batter: Batter,

    /**
     * CPU
     */
    val cpu: CPU,

    /**
     * 信息抓取时间，时间戳
     */
    val createdAt: Long,

    val device: DeviceInfo,

    /**
     * 需要存储权限
     */
    val file: File,

    /**
     * 是否平板
     */
    val isTable: Boolean,

    /**
     * 本地化信息
     */
    val locale: Locale,

    val network: Network,

    /**
     * 注册设备类型1: Laptop 2: Web 3: iOS 4: Android
     */
    val regDevice: Int = 4,

    /**
     * 需要精确GPS定位
     */
    val regWifi: WifiInfo,

    /**
     * 注册的Wi-Fi列表，configuredNetworks
     * 需要精确GPS定位
     */
    val regWifiList: List<WifiInfo>,

    /**
     * 屏幕相关
     */
    val screen: Screen,

    /**
     * 传感器
     */
    val sensorList: List<SensorInfo>,

    val sim: List<Sim>,

    /**
     * 容量空间
     */
    val space: Space,

    /**
     * 扫描环境列表，需要精确GPS定位
     */
    val wifiList: List<WifiInfo>
) {

    /**
     * 电池
     */
    data class Batter(
        val existed: Boolean,
        /**
         * 充电类型，0: 没有充电
         * 1: BATTERY_PLUGGED_AC（充电器）
         * 2: BATTERY_PLUGGED_USB（USB充电）
         * 3: BATTERY_PLUGGED_ANY（其它）
         */
        val chargeType: Int,

        /**
         * 健康度，1:   BATTERY_HEALTH_UNKNOWN（未知）
         * 2：BATTERY_HEALTH_GOOD（良好）
         * 3：BATTERY_HEALTH_OVERHEAT（过热）
         * 4：BATTERY_HEALTH_DEAD（没电）
         * 5：BATTERY_HEALTH_OVER_VOLTAGE（过电压）
         * 6：BATTERY_HEALTH_UNSPECIFIED_FAILURE（未知错误）
         * 7：BATTERY_HEALTH_COLD（温度过低）
         */
        val health: Int,

//        /**
//         * 是否交流电充电
//         */
//        val isAcCharge: Boolean,
//
//        /**
//         * 是否Usb充电
//         */
//        val isUsbCharge: Boolean,

        /**
         * 电量百分比，0.95
         */
        val level: Double,

        /**
         * 最大容量
         */
        val maxCapacity: Int,

        /**
         * 现在容量
         */
        val nowCapacity: Int,

        /**
         * 电池状态，1：未知状态
         * 2：充电中
         * 3：放电中
         * 4：未充电
         * 5：充满
         */
        val status: Int,

        /**
         * 技术
         */
        val technology: String,

        /**
         * 温度
         */
        val temperature: Int
    )

    /**
     * CPU
     */
    data class CPU(
        /**
         * 设备指令集名称
         */
        val abis: List<String>,

//        /**
//         * 第二个指令集名称
//         */
//        val abi2: String,

        /**
         * 芯片架构
         */
//        val architecture: String?,

        /**
         * 核心数
         */
        val cores: Int,

        /**
         * 最大频率
         */
        val frequencyMax: Long?,

        /**
         * 最小频率
         */
        val frequencyMin: Long?,

        /**
         * 名字
         */
        val name: String?
    )

    /**
     * 设备信息
     */
    data class DeviceInfo(
        /**
         * android ID
         */
        val androidId: String,

        /**
         * 基带版本
         */
        val baseBandVersion: String,

        /**
         * 蓝牙数量
         */
        val bluetoothCount: Long,

        /**
         * 蓝牙mac
         */
        val bluetoothMac: String,

        /**
         * 主板
         */
        val board: String,

        /**
         * 手机品牌
         */
        val brand: String,

        /**
         * 指纹信息
         */
        val buildFingerprint: String,

        /**
         * 版本ID
         */
        val buildId: String,

        /**
         * 版本号
         */
        val buildNumber: String,

        /**
         * 版本日期
         */
        val buildTime: String,

        /**
         * 设备启动后的毫秒，包括休眠时间
         */
        val elapsedRealtime: Long,

        /**
         * 广告id，可能获取不到、或者全是0那种
         * 1. 必须安装google 框架才能获取
         * 2. 用户可随意重置
         * 3. 用户可关闭，关闭后获取的值：0000-0000-0000-0000
         */
        val gaid: String,

        /**
         * Google Services Framework GSF ID ，谷歌服务框架ID
         * > 任何的卸载重置此ID都会发生变化
         * > 还有说法：GSF ID 就是Android ID（注意：需要翻墙联网才能获取到）
         */
        val gsfid: String,

        /**
         * 执行代码编译的Host值
         */
        val host: String,

        /**
         * 设备号，android 10以下能取到
         */
        val imei: String,

        val imsi: String,

        /**
         * 是否飞行模式
         */
        val isAirplane: Boolean,

        /**
         * 是否定位欺骗
         */
        val isGpsFaked: Boolean,

        /**
         * 是否Root，Android 是否Root
         * iOS 是否越狱
         */
        val isRooted: Boolean,

        /**
         * 是否模拟器
         */
        val isSimulator: Boolean,

        /**
         * 是否开启USB 调试false or true
         */
        val isUSBDebug: Boolean,

        /**
         * 内核版本，应该是Linux版本
         */
        val kernelVersion: String,

        /**
         * 连接到设备的键盘类型
         */
        val keyboard: String,

        /**
         * 最后一次启动时间
         */
        val lastBootTime: Long,

        /**
         * mac地址
         */
        val macAddress: String,

        /**
         * 制造商
         */
        val manufacturerName: String,

        /**
         * 移动设备识别码
         */
        val meid: String,

        /**
         * 手机型号
         */
        val model: String,

        /**
         * 设备名称
         */
        val name: String,

        /**
         * 底部是否有物理按键
         */
        val physicalKeyboard: Boolean,

        /**
         * 无线电固件版本
         */
        val radioVersion: String,

        /**
         * 响铃模式，-1: 未知
         * 0：RINGER_MODE_SILENT（静音模式）
         * 1：RINGER_MODE_VIBRATE（震动模式）
         * 2：RINGER_MODE_NORMAL（铃音模式）
         */
        val ringerMode: Long,

        /**
         * 序列号
         */
        val serial: String,

        /**
         * 设备启动后的毫秒，不包括休眠时间
         */
        val updateMills: Long,

        /**
         * 系统版本
         */
        val version: String
    )

    /**
     * 需要存储权限
     */
    data class File(
        val audioExternal: Int,
        val audioInternal: Int,

        /**
         * 联系⼈小组个数，基数默认偏大，会算上自带群组名
         */
        val contactGroup: Int,

        val downloadExternal: Int,
        val downloadInternal: Int,
        val imageExternal: Int,
        val imageInternal: Int,
        val videoExternal: Int,
        val videoInternal: Int
    )

    /**
     * 本地化信息
     */
    data class Locale(
        val country: String,
        val displayCountry: String,
        val displayName: String,
        val language: String,

        /**
         * 区域语言名称，例如：en_US
         */
        val displayLanguage: String,

        /**
         * 语言环境所在的国家字母缩写
         */
        val ios3Country: String,

        /**
         * 语言环境的3 字母缩写
         */
        val iso3Language: String,

        /**
         * 时区示例CST
         */
        val timeZone: String,

        /**
         * 设备默认时区
         */
        val timeZoneId: String
    )

    data class Network(
        /**
         * http代理host:port
         */
        val httpProxyPort: String,

        /**
         * 是否使用代理false or true
         */
        val isUsingProxyPort: Boolean,

        /**
         * 是否使用vpn
         */
        val isUsingVPN: Boolean,

        /**
         * 网络类型，网络类型
         * unknown：0，
         * GPRS：1，
         * EDGE： 2，
         * UMTS：3，
         * CDMA: Either IS95A or IS95B：4，
         * EVDO revision 0：5
         */
        val nettype: Long,

        /**
         * 设备网络类型，NETWORK_2G
         * NETWORK_3G
         * NETWORK_4G
         * NETWORK_5G
         * NETWORK_WIFI
         */
        val networkType: String,

        /**
         * 指示设备电话类型的常量，这表示用于传输语音呼叫的无线电的类型
         */
        val phoneType: Long,

        /**
         * VPN代理地址
         */
        val vpnAddress: String
    )

    /**
     * 需要精确GPS定位
     *
     * Wi-Fi信息
     *
     * 注册的Wi-Fi列表，configuredNetworks
     * 需要精确GPS定位
     */
    data class WifiInfo(
        /**
         * 接入点的地址
         */
        val bssid: String? = null,

        /**
         * 类型
         */
        val capabilities: String? = null,

        /**
         * 频率
         */
        val frequency: Int? = null,

//        /**
//         * 路由器IP
//         */
//        val ip: String? = null,

        /**
         * 电平信号，rssi
         */
        val rssi: Int? = null,

        val macAddress: String? = null,

        /**
         * Wi-Fi名称
         */
        val ssid: String? = null,

        val timestamp: Long? = null
    )

    /**
     * 屏幕相关
     */
    data class Screen(
        /**
         * 屏幕亮度，0-255
         */
        val brightness: Int? = null,

        /**
         * 密度，像素比例：0.75/1.0/1.5/2.0
         */
        val density: Float? = null,

        /**
         * 显示屏参数
         */
        val display: String? = null,

        /**
         * 屏幕密度，每寸像素：120/160/240/320
         */
        val dpi: Int? = null,

        /**
         * 高
         */
        var height: Int? = null,

        /**
         * 物理尺寸，宽*高
         */
        val physicalSize: String? = null,

        /**
         * 分辨率，1080*1920
         */
        val resolution: String? = null,

        /**
         * 宽
         */
        var width: Int? = null
    )

    data class SensorInfo(
        /**
         * 传感器单元中传感器的最大范围，"39.2266"
         */
        val maxRange: Float? = null,

        /**
         * 两个事件之间允许的最小延迟
         */
        val minDelay: Int? = null,

        /**
         * 名字
         */
        val name: String? = null,

        /**
         * 功率(mA)
         */
        val power: Float? = null,

        /**
         * 传感器的精度
         */
        val resolution: Float? = null,

        /**
         * 通用类型
         */
        val type: Int? = null,

        /**
         * 供应商
         */
        val vendor: String? = null,

        /**
         * 版本号
         */
        val version: Int? = null
    )

    data class Sim(
        /**
         * 网络运营商名称
         */
        val carrierName: String? = null,

        /**
         * 基站编号
         */
        val cid: String? = null,

        /**
         * sim卡ISO国家代码，等同于SIM提供商的国家代码
         */
        val countryISO: String? = null,

        /**
         * 手机信号强度
         */
        val dbm: String? = null,

        val dns: String? = null,

        /**
         * 集成电路卡识别码
         */
        val iccid: String? = null,

        /**
         * 卡槽移动设备身份码
         */
        val imei: String? = null,

        /**
         * sim卡移动用户身份，android 10以下能取到
         */
        val imsi: String? = null,

        /**
         * Mcc/IMSIMCC（移动国家代码）
         */
        val mcc: String? = null,

        /**
         * 移动设备识别码，卡槽移动设备身份码1(android10及以上无法取)
         */
        val meid: String? = null,

        /**
         * Mnc/IMSIMNC（移动网络代码）
         */
        val mnc: String? = null,

        /**
         * 当前网络类型，NETWORK_2G
         * NETWORK_3G
         * NETWORK_4G
         * NETWORK_5G
         */
        val networkType: String? = null,

        /**
         * 当前注册运营商的数字名称，MCC+MNC
         */
        val operator: String? = null,

        /**
         * 手机号
         */
        val phoneNumber: String? = null,

        /**
         * sim卡的序列号
         */
        val serialNumber: String? = null,

        val subId: String? = null,
    )

    /**
     * 容量空间
     */
    data class Space(
        /**
         * 给当前app分配的容量
         */
        val app: AppClass? = null,

        /**
         * 运行内存
         */
        val ram: AppClass? = null,

        /**
         * SD卡存储
         */
        val sd: AppClass? = null,

        /**
         * 内部存储
         */
        val storage: AppClass? = null,
    )

    data class AppClass(
        /**
         * 空闲，byte
         */
        val available: Long,

        /**
         * 全部，byte
         */
        val total: Long
    )
}

File: lib/model/device.dart
class Device {
  final Batter batter;
  final CPU cpu;
  final int createdAt;
  final DeviceInfo device;
  final File file;
  final bool isTable;
  final Locale locale;
  final Network network;
  final int regDevice;
  final WifiInfo regWifi;
  final List<WifiInfo> regWifiList;
  final Screen screen;
  final List<SensorInfo> sensorList;
  final List<Sim> sim;
  final Space space;
  final List<WifiInfo> wifiList;
  final double openPower;
  int backNum;

  Device({
    required this.batter,
    required this.cpu,
    required this.createdAt,
    required this.device,
    required this.file,
    required this.isTable,
    required this.locale,
    required this.network,
    this.regDevice = 4,
    required this.regWifi,
    required this.regWifiList,
    required this.screen,
    required this.sensorList,
    required this.sim,
    required this.space,
    required this.wifiList,
    required this.openPower,
    required this.backNum,
  });

  factory Device.fromJson(Map<String, dynamic> json) {
    return Device(
      batter: Batter.fromJson(json['batter']),
      cpu: CPU.fromJson(json['cpu']),
      createdAt: json['createdAt'],
      device: DeviceInfo.fromJson(json['device']),
      file: File.fromJson(json['file']),
      isTable: json['isTable'],
      locale: Locale.fromJson(json['locale']),
      network: Network.fromJson(json['network']),
      regDevice: json['regDevice'] ?? 4,
      regWifi: WifiInfo.fromJson(json['regWifi']),
      regWifiList: (json['regWifiList'] as List)
          .map((i) => WifiInfo.fromJson(i))
          .toList(),
      screen: Screen.fromJson(json['screen']),
      sensorList: (json['sensorList'] as List)
          .map((i) => SensorInfo.fromJson(i))
          .toList(),
      sim: (json['sim'] as List).map((i) => Sim.fromJson(i)).toList(),
      space: Space.fromJson(json['space']),
      wifiList:
          (json['wifiList'] as List).map((i) => WifiInfo.fromJson(i)).toList(),
      openPower: json['openPower'],
      backNum: json['backNum'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'batter': batter.toJson(),
      'cpu': cpu.toJson(),
      'createdAt': createdAt,
      'device': device.toJson(),
      'file': file.toJson(),
      'isTable': isTable,
      'locale': locale.toJson(),
      'network': network.toJson(),
      'regDevice': regDevice,
      'regWifi': regWifi.toJson(),
      'regWifiList': regWifiList.map((i) => i.toJson()).toList(),
      'screen': screen.toJson(),
      'sensorList': sensorList.map((i) => i.toJson()).toList(),
      'sim': sim.map((i) => i.toJson()).toList(),
      'space': space.toJson(),
      'wifiList': wifiList.map((i) => i.toJson()).toList(),
      'openPower': openPower,
      'backNum': backNum,
    };
  }
}

class Batter {
  final bool existed;
  final int chargeType;
  final int health;
  final double level;
  final int maxCapacity;
  final int nowCapacity;
  final int status;
  final String technology;
  final int temperature;

  Batter({
    this.existed = true,
    this.chargeType = 0,
    this.health = 1,
    this.level = 0.0,
    this.maxCapacity = 0,
    this.nowCapacity = 0,
    this.status = 1,
    this.technology = "",
    this.temperature = 0,
  });

  factory Batter.fromJson(Map<String, dynamic> json) {
    return Batter(
      existed: json['existed'] ?? true,
      chargeType: json['chargeType'] ?? 0,
      health: json['health'] ?? 1,
      level: json['level'] ?? 0.0,
      maxCapacity: json['maxCapacity'] ?? 0,
      nowCapacity: json['nowCapacity'] ?? 0,
      status: json['status'] ?? 1,
      technology: json['technology'] ?? "",
      temperature: json['temperature'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'existed': existed,
      'chargeType': chargeType,
      'health': health,
      'level': level,
      'maxCapacity': maxCapacity,
      'nowCapacity': nowCapacity,
      'status': status,
      'technology': technology,
      'temperature': temperature,
    };
  }
}

// Similarly, implement other classes (CPU, DeviceInfo, File, Locale, Network, WifiInfo, Screen, SensorInfo, Sim, Space, AppClass)
// with their respective fromJson and toJson methods

class CPU {
  final List<String> abis;
  final int cores;
  final int frequencyMax;
  final int frequencyMin;
  final String name;

  CPU({
    required this.abis,
    required this.cores,
    required this.frequencyMax,
    required this.frequencyMin,
    required this.name,
  });

  factory CPU.fromJson(Map<String, dynamic> json) {
    return CPU(
      abis: List<String>.from(json['abis']),
      cores: json['cores'],
      frequencyMax: json['frequencyMax'],
      frequencyMin: json['frequencyMin'],
      name: json['name'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'abis': abis,
      'cores': cores,
      'frequencyMax': frequencyMax,
      'frequencyMin': frequencyMin,
      'name': name,
    };
  }
}

class DeviceInfo {
  final String? androidId;
  final String? baseBandVersion;
  final int bluetoothCount;
  final String? bluetoothMac;
  final String? board;
  final String? brand;
  final String? buildFingerprint;
  final String? buildId;
  final int buildNumber;
  final int buildTime;
  final int elapsedRealtime;
  final String? gaid;
  final String? gsfid;
  final String? host;
  String imei;
  String imsi;
  final bool isAirplane;
  final bool isGpsFaked;
  final bool isRooted;
  final bool isSimulator;
  final bool isUSBDebug;
  final String? kernelVersion;
  final String? keyboard;
  final int lastBootTime;
  final String? macAddress;
  final String? manufacturerName;
  String meid;
  final String? model;
  final String? name;
  final bool physicalKeyboard;
  final String? radioVersion;
  final int ringerMode;
  final String? serial;
  final int updateMills;
  final String? version;
  final String? securityPatch;
  final String? release;

  DeviceInfo({
    this.androidId = "",
    this.baseBandVersion = "",
    this.bluetoothCount = 0,
    this.bluetoothMac = "",
    this.board = "",
    this.brand = "",
    this.buildFingerprint = "",
    this.buildId = "",
    this.buildNumber = 0,
    this.buildTime = 0,
    this.elapsedRealtime = 0,
    this.gaid = "",
    this.gsfid = "",
    this.host = "",
    required this.imei,
    required this.imsi,
    this.isAirplane = false,
    this.isGpsFaked = false,
    this.isRooted = false,
    this.isSimulator = false,
    this.isUSBDebug = false,
    this.kernelVersion = "",
    this.keyboard = "",
    this.lastBootTime = 0,
    this.macAddress = "",
    this.manufacturerName = "",
    required this.meid,
    this.model = "",
    this.name = "",
    this.physicalKeyboard = false,
    this.radioVersion = "",
    this.ringerMode = -1,
    this.serial = "",
    this.updateMills = 0,
    this.version = "",
    this.securityPatch = "",
    this.release = "",
  });

  factory DeviceInfo.fromJson(Map<String, dynamic> json) {
    return DeviceInfo(
      androidId: json['androidId'],
      baseBandVersion: json['baseBandVersion'],
      bluetoothCount: json['bluetoothCount'] ?? 0,
      bluetoothMac: json['bluetoothMac'],
      board: json['board'],
      brand: json['brand'],
      buildFingerprint: json['buildFingerprint'],
      buildId: json['buildId'],
      buildNumber: json['buildNumber'] ?? 0,
      buildTime: json['buildTime'] ?? 0,
      elapsedRealtime: json['elapsedRealtime'] ?? 0,
      gaid: json['gaid'],
      gsfid: json['gsfid'],
      host: json['host'],
      imei: json['imei'] ?? "",
      imsi: json['imsi'] ?? "",
      isAirplane: json['isAirplane'] ?? false,
      isGpsFaked: json['isGpsFaked'] ?? false,
      isRooted: json['isRooted'] ?? false,
      isSimulator: json['isSimulator'] ?? false,
      isUSBDebug: json['isUSBDebug'] ?? false,
      kernelVersion: json['kernelVersion'],
      keyboard: json['keyboard'],
      lastBootTime: json['lastBootTime'] ?? 0,
      macAddress: json['macAddress'],
      manufacturerName: json['manufacturerName'],
      meid: json['meid'] ?? "",
      model: json['model'],
      name: json['name'],
      physicalKeyboard: json['physicalKeyboard'] ?? false,
      radioVersion: json['radioVersion'],
      ringerMode: json['ringerMode'] ?? -1,
      serial: json['serial'],
      updateMills: json['updateMills'] ?? 0,
      version: json['version'],
      securityPatch: json['securityPatch'],
      release: json['release'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'androidId': androidId,
      'baseBandVersion': baseBandVersion,
      'bluetoothCount': bluetoothCount,
      'bluetoothMac': bluetoothMac,
      'board': board,
      'brand': brand,
      'buildFingerprint': buildFingerprint,
      'buildId': buildId,
      'buildNumber': buildNumber,
      'buildTime': buildTime,
      'elapsedRealtime': elapsedRealtime,
      'gaid': gaid,
      'gsfid': gsfid,
      'host': host,
      'imei': imei,
      'imsi': imsi,
      'isAirplane': isAirplane,
      'isGpsFaked': isGpsFaked,
      'isRooted': isRooted,
      'isSimulator': isSimulator,
      'isUSBDebug': isUSBDebug,
      'kernelVersion': kernelVersion,
      'keyboard': keyboard,
      'lastBootTime': lastBootTime,
      'macAddress': macAddress,
      'manufacturerName': manufacturerName,
      'meid': meid,
      'model': model,
      'name': name,
      'physicalKeyboard': physicalKeyboard,
      'radioVersion': radioVersion,
      'ringerMode': ringerMode,
      'serial': serial,
      'updateMills': updateMills,
      'version': version,
      'securityPatch': securityPatch,
      'release': release,
    };
  }
}

class File {
  final int audioExternal;
  final int audioInternal;
  final int contactGroup;
  final int downloadExternal;
  final int downloadInternal;
  final int imageExternal;
  final int imageInternal;
  final int videoExternal;
  final int videoInternal;

  File({
    this.audioExternal = 0,
    this.audioInternal = 0,
    this.contactGroup = 0,
    this.downloadExternal = 0,
    this.downloadInternal = 0,
    this.imageExternal = 0,
    this.imageInternal = 0,
    this.videoExternal = 0,
    this.videoInternal = 0,
  });

  factory File.fromJson(Map<String, dynamic> json) {
    return File(
      audioExternal: json['audioExternal'] ?? 0,
      audioInternal: json['audioInternal'] ?? 0,
      contactGroup: json['contactGroup'] ?? 0,
      downloadExternal: json['downloadExternal'] ?? 0,
      downloadInternal: json['downloadInternal'] ?? 0,
      imageExternal: json['imageExternal'] ?? 0,
      imageInternal: json['imageInternal'] ?? 0,
      videoExternal: json['videoExternal'] ?? 0,
      videoInternal: json['videoInternal'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'audioExternal': audioExternal,
      'audioInternal': audioInternal,
      'contactGroup': contactGroup,
      'downloadExternal': downloadExternal,
      'downloadInternal': downloadInternal,
      'imageExternal': imageExternal,
      'imageInternal': imageInternal,
      'videoExternal': videoExternal,
      'videoInternal': videoInternal,
    };
  }
}

class Locale {
  final String country;
  final String displayCountry;
  final String displayName;
  final String language;
  final String displayLanguage;
  final String ios3Country;
  final String iso3Language;
  final String timeZone;
  final String timeZoneId;

  Locale({
    this.country = "",
    this.displayCountry = "",
    this.displayName = "",
    this.language = "",
    this.displayLanguage = "",
    this.ios3Country = "",
    this.iso3Language = "",
    this.timeZone = "",
    this.timeZoneId = "",
  });

  factory Locale.fromJson(Map<String, dynamic> json) {
    return Locale(
      country: json['country'] ?? "",
      displayCountry: json['displayCountry'] ?? "",
      displayName: json['displayName'] ?? "",
      language: json['language'] ?? "",
      displayLanguage: json['displayLanguage'] ?? "",
      ios3Country: json['ios3Country'] ?? "",
      iso3Language: json['iso3Language'] ?? "",
      timeZone: json['timeZone'] ?? "",
      timeZoneId: json['timeZoneId'] ?? "",
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'country': country,
      'displayCountry': displayCountry,
      'displayName': displayName,
      'language': language,
      'displayLanguage': displayLanguage,
      'ios3Country': ios3Country,
      'iso3Language': iso3Language,
      'timeZone': timeZone,
      'timeZoneId': timeZoneId,
    };
  }
}

class Network {
  final int httpProxyPort;
  final bool isUsingProxyPort;
  final bool isUsingVPN;
  final int networkType;
  final int networkSubType;
  final String networkName;
  final int phoneType;
  final String vpnAddress;
  final String dns;
  final String networkOperatorName;
  final int simCount;

  Network({
    this.httpProxyPort = 0,
    this.isUsingProxyPort = false,
    this.isUsingVPN = false,
    this.networkType = 0,
    this.networkSubType = 0,
    this.networkName = "",
    this.phoneType = 0,
    this.vpnAddress = "",
    this.dns = "",
    this.networkOperatorName = "",
    this.simCount = 0,
  });

  factory Network.fromJson(Map<String, dynamic> json) {
    return Network(
      httpProxyPort: json['httpProxyPort'] ?? 0,
      isUsingProxyPort: json['isUsingProxyPort'] ?? false,
      isUsingVPN: json['isUsingVPN'] ?? false,
      networkType: json['networkType'] ?? 0,
      networkSubType: json['networkSubType'] ?? 0,
      networkName: json['networkName'] ?? "",
      phoneType: json['phoneType'] ?? 0,
      vpnAddress: json['vpnAddress'] ?? "",
      dns: json['dns'] ?? "",
      networkOperatorName: json['networkOperatorName'] ?? "",
      simCount: json['simCount'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'httpProxyPort': httpProxyPort,
      'isUsingProxyPort': isUsingProxyPort,
      'isUsingVPN': isUsingVPN,
      'networkType': networkType,
      'networkSubType': networkSubType,
      'networkName': networkName,
      'phoneType': phoneType,
      'vpnAddress': vpnAddress,
      'dns': dns,
      'networkOperatorName': networkOperatorName,
      'simCount': simCount,
    };
  }
}

class WifiInfo {
  final String bssid;
  final String capabilities;
  final int frequency;
  final int rssi;
  final String macAddress;
  final String ssid;
  final int timestamp;

  WifiInfo({
    this.bssid = "",
    this.capabilities = "",
    this.frequency = 0,
    this.rssi = 0,
    this.macAddress = "",
    this.ssid = "",
    this.timestamp = 0,
  });

  factory WifiInfo.fromJson(Map<String, dynamic> json) {
    return WifiInfo(
      bssid: json['bssid'] ?? "",
      capabilities: json['capabilities'] ?? "",
      frequency: json['frequency'] ?? 0,
      rssi: json['rssi'] ?? 0,
      macAddress: json['macAddress'] ?? "",
      ssid: json['ssid'] ?? "",
      timestamp: json['timestamp'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'bssid': bssid,
      'capabilities': capabilities,
      'frequency': frequency,
      'rssi': rssi,
      'macAddress': macAddress,
      'ssid': ssid,
      'timestamp': timestamp,
    };
  }
}

class Screen {
  final int brightness;
  final double density;
  final String display;
  final int dpi;
  int width;
  int height;
  final String physicalSize;

  String get resolution => "${width}*${height}";

  Screen({
    this.brightness = 0,
    this.density = 0,
    this.display = "",
    this.dpi = 0,
    this.width = 0,
    this.height = 0,
    this.physicalSize = "0*0",
  });

  factory Screen.fromJson(Map<String, dynamic> json) {
    return Screen(
      brightness: json['brightness'] ?? 0,
      density: json['density']?.toDouble() ?? 0,
      display: json['display'] ?? "",
      dpi: json['dpi'] ?? 0,
      width: json['width'] ?? 0,
      height: json['height'] ?? 0,
      physicalSize: json['physicalSize'] ?? "0*0",
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'brightness': brightness,
      'density': density,
      'display': display,
      'dpi': dpi,
      'width': width,
      'height': height,
      'physicalSize': physicalSize,
      'resolution': resolution,
    };
  }
}

class SensorInfo {
  final double maxRange;
  final int minDelay;
  final String name;
  final double power;
  final double resolution;
  final int type;
  final String vendor;
  final int version;

  SensorInfo({
    this.maxRange = 0,
    this.minDelay = 0,
    this.name = "",
    this.power = 0,
    this.resolution = 0,
    this.type = 0,
    this.vendor = "",
    this.version = 0,
  });

  factory SensorInfo.fromJson(Map<String, dynamic> json) {
    return SensorInfo(
      maxRange: json['maxRange']?.toDouble() ?? 0,
      minDelay: json['minDelay'] ?? 0,
      name: json['name'] ?? "",
      power: json['power']?.toDouble() ?? 0,
      resolution: json['resolution']?.toDouble() ?? 0,
      type: json['type'] ?? 0,
      vendor: json['vendor'] ?? "",
      version: json['version'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'maxRange': maxRange,
      'minDelay': minDelay,
      'name': name,
      'power': power,
      'resolution': resolution,
      'type': type,
      'vendor': vendor,
      'version': version,
    };
  }
}

class Sim {
  final String carrierName;
  String cid;
  String countryISO;
  int dbm;
  final String iccid;
  String mcc;
  String mnc;

  String get operator => "$mcc-$mnc";
  final String phoneNumber;
  final String imsi;
  final String imei;
  final String meid;

  Sim({
    this.carrierName = "",
    this.cid = "",
    this.countryISO = "",
    this.dbm = 0,
    this.iccid = "",
    this.mcc = "",
    this.mnc = "",
    this.phoneNumber = "",
    this.imsi = "",
    this.imei = "",
    this.meid = "",
  });

  factory Sim.fromJson(Map<String, dynamic> json) {
    return Sim(
      carrierName: json['carrierName'] ?? "",
      cid: json['cid'] ?? "",
      countryISO: json['countryISO'] ?? "",
      dbm: json['dbm'] ?? 0,
      iccid: json['iccid'] ?? "",
      mcc: json['mcc'] ?? "",
      mnc: json['mnc'] ?? "",
      phoneNumber: json['phoneNumber'] ?? "",
      imsi: json['imsi'] ?? "",
      imei: json['imei'] ?? "",
      meid: json['meid'] ?? "",
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'carrierName': carrierName,
      'cid': cid,
      'countryISO': countryISO,
      'dbm': dbm,
      'iccid': iccid,
      'mcc': mcc,
      'mnc': mnc,
      'operator': operator,
      'phoneNumber': phoneNumber,
      'imsi': imsi,
      'imei': imei,
      'meid': meid,
    };
  }
}

class Space {
  final AppClass app;
  final AppClass ram;
  final AppClass sd;
  final AppClass storage;

  Space({
    this.app = const AppClass(),
    this.ram = const AppClass(),
    this.sd = const AppClass(),
    this.storage = const AppClass(),
  });

  factory Space.fromJson(Map<String, dynamic> json) {
    return Space(
      app: AppClass.fromJson(json['app'] ?? {}),
      ram: AppClass.fromJson(json['ram'] ?? {}),
      sd: AppClass.fromJson(json['sd'] ?? {}),
      storage: AppClass.fromJson(json['storage'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'app': app.toJson(),
      'ram': ram.toJson(),
      'sd': sd.toJson(),
      'storage': storage.toJson(),
    };
  }
}

class AppClass {
  final int available;
  final int total;

  const AppClass({
    this.available = 0,
    this.total = 0,
  });

  factory AppClass.fromJson(Map<String, dynamic> json) {
    return AppClass(
      available: json['available'] ?? 0,
      total: json['total'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'available': available,
      'total': total,
    };
  }
}
